Deprecate usage of Rack::Runtime in the default middleware stack reso…

SValkanov opened the pull request:
Deprecate usage of Rack::Runtime in the default middleware stack
resolves #38325

SValkanov committed:
Deprecate usage of Rack::Runtime in the default middleware stack resolves #38325

kevinnio commented:
This is good for deprecation warnings but what about removing the `X-RUNTIME` response header mentioned in the original issue?

SValkanov commented:
@kevinnio, As @tenderlove mentioned, people are depending that `Rack::Runtime` is included in the default middleware stack. So I approached the situation gradually. First raising deprecation notices that it will be removed which will give people time to migrate, and the next step is actually removing it in the given future version of Rails.

tenderlove commented:
I'm not sure this is what we want to do.  Won't this just cause warnings on every Rails application unless the users specifically remove the middleware?  The issue isn't people specifically adding the Rack::Runtime middleware, it's that people may have code like this:

```ruby
config.middleware.insert_before(::Rack::Runtime, SomeCoolMiddleware.new)
```

The user never asked for `Rack::Runtime` to be added, but they want their "cool middleware" to be inserted *before* `Rack::Runtime`.  If `Rack::Runtime` doesn't exist in the middleware stack, then the above code won't work.

I think the best way to go about fixing this is probably to insert a dummy middleware, then check for people passing in `Rack::Runtime` to any methods that manipulate the middleware stack.

SValkanov commented:
@tenderlove I'm not sure that I understand correctly the following: 

> I think the best way to go about fixing this is probably to insert a dummy middleware

Do you mean to extend `Rack::Runtime` in some way, so it will be still present in the middleware, but to do not include `X-RUNTIME` in the response header?

Also is it a good idea to patch the methods of `MiddlewareStackProxy` dynamically to check if `Rack::Runtime` is passed?

SValkanov committed:
Rework of deprecating `Rack::Runtime` in the default middleware stack

Added checks if `Rack::Runtime` is passed to any methods that manipulate the middleware stack - `MiddlewareStackProxy`

Changed initialization of cache middleware
